enhancements/vdpa/virtio-vdpa-ovn-hw-offload.md

```diff
@@ -0,0 +1,235 @@
+
+# Virtio/vDPA with OVN HW offload
+
+The purpose of this enhancement is to provide a proposal for integrating Virtio/vDPA in OCP for container workloads.
+
+
+## Summary
+See [VDPA support in OCP (Overview)](https://github.com/openshift/enhancements/blob/master/enhancements/vdpa/vdpa-support-overview.md).
+
+Main aspects of this enhancement:
+- primary interface in the pod is configured as a Virtio/vDPA device
+- NIC configured in switchdev mode
+- OVS HW offloading enabled
+- Supported HW: NVIDIA CTX-6 DX
+
+
+## Motivation
+
+The main motivation is to integrate a Virtio/vDPA device in Openshift.
+
+
+### User Stories
+
+Details are provided in the overview enhancement.
+
+### Goals
+
+- As a user I want to configure an Openshift cluster with Virtio/VDPA support.
+- As a user I want to determine how many VFs to create for a given NIC and how each VF should be consumed by workloads (e.g., netdev device in the container network namespace)
+- As a user I want the VDPA solution to be integrated with OVN-kubernetes and with HW offloading enabled
+
+### Non-Goals
+
+- The Vhost-vdpa driver won't be supported in CNFs (VMs only). The virtio-vdpa driver will be supported in CNFs, i.e. netdev device to be consumed.
+- Scalable IOV and Sub-functions are out of scope, only switchdev SR-IOV will be supported (with OVS HW offloading to the NIC)
+
+## Proposal
+
+The following diagram depicts all the involved components and their interactions:
+
+![alt text](vdpa-architecture.png "VDPA architecture")
+
+### Workflow Description
+
+#### Pre-requirements
+
+- Create the Openshift cluster on a baremetal node
+- Deploy OVN-kubernetes
+- Deploy the SRIOV network operator
+
+#### SRIOV network operator workflow
+
+- Deploy the SRIOV device plugin
+- Deploy OpenVswitch in HW offload mode
+- Configure the VFs (same as SR-IOV: echo N > /sys/devices/{PF}/sriov_numvfs)
+- Install vendor-independent kernel drivers: vdpa, virtio-vdpa
+- Install vendor-dependent kernel drivers, e.g. mlx5-vdpa for Mellanox cards
+- Put NIC in switchdev mode (smart NIC)
+- (Vendor specific) Bind the right PCI driver. Some vendors might implement vdpa in a different PCI driver (e.g: Intel’s ifcvf). Others might keep the same pci driver and require extra steps (e.g: in mellanox vdpa, the VF is still bound to “mlx5_core”)
+- Enable HW offload mode on PF, VFs and port representors
+- (Vendor specific) Create the vdpa device. Some vendors might require no extra steps because they create the vdpa device on a pci driver probe (e.g: ifcvf). Others might need extra steps (e.g: Mellanox requires loading an additional driver and in the future, they 	might require managing a “virtual bus”(source)). The plan is to extend the govdpa library to provide such functionality.
+- Bind the vdpa device to the correct vdpa bus driver (virtio-vdpa driver in the first implementation). This is not vendor specific and uses a sysfs-based API.
+- Add the appropriate vdpa flags to the SR-IOV Device Plugin’s configMap.
```

Comment:
I think this changes need to be propagated from the sriov operator maybe some changes are needed to the sriovNetworkNodePolicy object

Reply:
Yes, the changes to the policy are mentioned in the API extension paragraph. Anything else am I missing?